staging/src/k8s.io/apiserver/pkg/server/filters/compression.go

```diff
@@ -0,0 +1,147 @@
+			compressionWriter.Header().Set("Content-Encoding", encoding)
+			handler.ServeHTTP(compressionWriter, req)
+		} else {
+			handler.ServeHTTP(w, req)
+		}
+	})
+
+}
+
+// WantsCompressedResponse reads the Accept-Encoding header to see if and which encoding is requested.
+func wantsCompressedResponse(httpRequest *http.Request) (bool, string) {
+	if watch := httpRequest.URL.Query().Get("watch"); watch == "true" || watch == "1" {
+		return false, ""
+	}
+	header := httpRequest.Header.Get(header_AcceptEncoding)
+	gi := strings.Index(header, encoding_gzip)
+	zi := strings.Index(header, encoding_deflate)
+	// use in order of appearance
+	if gi == -1 {
+		return zi != -1, encoding_deflate
+	} else if zi == -1 {
+		return gi != -1, encoding_gzip
+	} else {
+		if gi < zi {
+			return true, encoding_gzip
+		}
+		return true, encoding_deflate
+	}
+}
+
+type compressionResponseWriter struct {
+	writer     http.ResponseWriter
+	compressor Compressor
+	encoding   string
+}
+
+func NewCompressionResponseWriter(w http.ResponseWriter, encoding string) *compressionResponseWriter {
+	var compressor Compressor
+	switch encoding {
+	case encoding_gzip:
+		compressor = gzip.NewWriter(w)
+	case encoding_deflate:
+		compressor = zlib.NewWriter(w)
+	default:
+		panic(encoding + " not a valid compression type")
+	}
+	return &compressionResponseWriter{
+		writer:     w,
+		compressor: compressor,
+		encoding:   encoding,
+	}
+}
+
+// compressionResponseWriter implements http.Responsewriter Interface
+var _ http.ResponseWriter = &compressionResponseWriter{}
+
+func (c *compressionResponseWriter) Header() http.Header {
+	return c.writer.Header()
+}
+
+// compress data according to compression method
+func (c *compressionResponseWriter) Write(p []byte) (int, error) {
+	if c.isCompressorClosed() {
+		return -1, errors.New("Compressing error: tried to write data using closed compressor")
+	}
+	c.Header().Set(header_ContentEncoding, c.encoding)
+	//glog.V(0).Infof("WROTE \n\n%s\n\nwith headers: %+v\n", p, c.Header())
+	defer c.Close()
+	return c.compressor.Write(p)
+}
+
+func (c *compressionResponseWriter) WriteHeader(status int) {
+	c.writer.WriteHeader(status)
+}
+
+// CloseNotify is part of http.CloseNotifier interface
+func (c *compressionResponseWriter) CloseNotify() <-chan bool {
+	return c.writer.(http.CloseNotifier).CloseNotify()
+}
+
+// Close the underlying compressor
+func (c *compressionResponseWriter) Close() error {
+	if c.isCompressorClosed() {
+		return errors.New("Compressing error: tried to close already closed compressor")
+	}
+
+	c.compressor.Close()
+	c.compressor = nil
+	return nil
+}
+
+func (c *compressionResponseWriter) Flush() {
+	c.compressor.Flush()
+}
+
+func (c *compressionResponseWriter) isCompressorClosed() bool {
+	return nil == c.compressor
+}
+
+func RestfulWithCompression(function restful.RouteFunction) restful.RouteFunction {
+	return restful.RouteFunction(func(request *restful.Request, response *restful.Response) {
```

Comment:
Does the following seem right to you? I feel like I'm messing with `restful` internals here, but this was how I implemented the delegation:
```go
func RestfulWithCompression(function restful.RouteFunction) restful.RouteFunction {
	return restful.RouteFunction(func(request *restful.Request, response *restful.Response) {
		handler := WithCompression(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
			response.ResponseWriter = w
			request.Request = req
			function(request, response)
		}))
		handler.ServeHTTP(response.ResponseWriter, request.Request)
	})
}
```
I didn't want to go with this as I was more comfortable being explicit with the code (despite their being code duplication)